### PR TITLE
Improve performance of Random.sample when use_true_random=False

### PR DESCRIPTION
**Pull request by DRMacIver**
This improves the performance of sampling mostly by making it fail when the stream is being manipulated in a way that works against it, which happens e.g. when passing all zeroes to a non-zero sample, and can trigger a health check.

This was prompted by #2503 failing due to flaky coverage of this method, and noticing this problem when I went to try to write more precise tests for it. This yak stack is so deep right now...

### CODE REVIEW

**Comment by Zalathar** [comment]
Something weird must have happened, because this PR currently contains no code.

**Comment by DRMacIver** [comment]
> Something weird must have happened, because this PR currently contains no code.

Huh. That *is* weird. Let me try again.

**Comment by DRMacIver** [comment]
And it's also completely gone from my local history. How incredibly irritating.

**Comment by DRMacIver** [comment]
Fortunately the code in question was not that complicated so I've recreated it from memory.

**Comment by Zac-HD** [comment]
Huh, I'd noticed the same thing in #2506 but thought that `UniqueSampledListStrategy` was a better solution, since the linear (rather than quadratic) worst-case behaviour helps a lot for `n`-choose-`k~n` samples. 
 Branch here https://github.com/HypothesisWorks/hypothesis/compare/master...Zac-HD:linear-sample, though I think we'd still need your version here to avoid problems on the very-large-population path.

**Comment by DRMacIver** [comment]
> since the linear (rather than quadratic) worst-case behaviour 

The worst case behaviour isn't actually quadratic I think, it's a [coupon collector problem](https://en.wikipedia.org/wiki/Coupon_collector%27s_problem) so this approach is `O(log(n))` when running in pure random mode.

**Comment by Zac-HD** [comment]
Sorry, you're right, the *expected* time to generate n=k unique elements in true random mode is O nlogn; though the worst case is unbounded.

Assuming that our caching stops us from retrying bad elements though, the worst-case is still quadratic because we can try n-1 wrong guesses for the last element, n-2 for the one before that, and so on.

In any case, O(n) time to choose n elements is still better!  And come to think of it, if we `LazySequenceCopy` the `range()` object we don't need to worry about memory use for extremely large sequences either.

**Comment by DRMacIver** [comment]
Fair points, all. I confess I had forgotten that we had that special casing already available, and I've updated the code to use it.

I don't think it's valid to `LazySequenceCopy` in `sampled_from` in general because the sequence we're copying might actually be mutable - it's a protection against the caller mutating it, not because we want to do so, but we can at least special case `range`.

**Comment by Zac-HD** [comment]
Yeah, I was mostly concerned with the behaviour if you call `Random().sample(population=range(10**100), k=100)`... this works fine with the stdlib code, but could easily explode with our backend.

I was thinking actually of the `range` object constructed in https://github.com/HypothesisWorks/hypothesis/compare/master...Zac-HD:linear-sample for the indices rather than the user-supplied sequence.  That branch shows that we'll need to special-case `range` in `st.sampled_from` too for this to work, though, and ideally add a test using `st.lists(st.sampled_from(range(10**100)), min_size=10, unique=True)` to check for blowups.

**Comment by DRMacIver** [comment]
> That branch shows that we'll need to special-case range in st.sampled_from too for this to work

Does it? It shows that we need to special case it in the repr for `SampledFromStrategy`, but also looking at that code shows that the repr for `SampledFromStrategy` was entirely doing the wrong thing anyway. 😢

**Comment by Zac-HD** [comment]
Yes, because we'll cast it to a `tuple` in `check_sample` otherwise.

**Comment by DRMacIver** [comment]
> Yes, because we'll cast it to a tuple in check_sample otherwise.

Right. But it already needs a special case in `check_sample`, so why does it need one in `sampled_from`?

Also it turns out that the sampled_from repr is never getting used and that there's a giant mess of yuck about how our lazy strategy initialisation interacts with mutability that I don't want to get into right now. 😢

**Comment by Zac-HD** [comment]
Right, I meant that we need the special case in `check_sample` for `sampled_from(range(10**100))` to work.

:scream: yeah, that doesn't look fun.

**Comment by DRMacIver** [comment]
Oops apparently I pushed the change to the wrong branch earlier. I've pushed a bit that uses the `lists(sampled_from(...))` approach now. Not going to touch the rest of this.

**Comment by Zac-HD** [comment]
> Oops apparently I pushed the change to the wrong branch earlier.

Relatable... I've had a couple lately where I ammended and force-pushed *in the wrong repo* which is :grimacing:
